File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.java
class Solution {
    public int maximumCandies(int[] candies, long k) {
        long total = Arrays.stream(candies).asLongStream().sum();
        if(total<k) return 0;
        int left = 1;
        int ryt = Arrays.stream(candies).max().getAsInt();
        while(left < ryt){
            int mid = (left+ryt+1)/2;
            long count = 0;
            for(int candy : candies) count += (candy/mid);
            if(count>= k) left = mid;
            else ryt = mid-1;
        }
        return left;
    }
}

File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.py
from typing import List

class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        if sum(candies) < k:
            return 0  # If total candies are less than children, each child gets 0.

        left, right = 1, max(candies)  # Binary search range

        while left < right:
            mid = (left + right + 1) // 2  # Middle value for binary search
            count = sum(c // mid for c in candies)  # Number of children that can be served

            if count >= k:  # If we can serve at least k children, try for more
                left = mid
            else:  # Otherwise, reduce the candy count per child
                right = mid - 1

        return left  # Maximum number of candies each child can get
